libs/extensions/src/NanoEvent.cpp

```diff
@@ -41,53 +41,56 @@ shared_ptr<PhysicsObjects> NanoEvent::GetOuterDRMatchedMuons(float matchingDelta
+    float phi1 = dsaMuon->Get("outerPhi");
+    bool matchFound = false;
+    for(auto muon : *looseMuons){
+      float eta2 = muon->Get("outerEta");
+      float phi2 = muon->Get("outerPhi");
+      if(DeltaR(eta1, phi1, eta2, phi2) < matchingDeltaR) matchFound = true;
     }
+    if(matchFound == false) allMuons->push_back(dsaMuon);
   }
 
   return allMuons;
 }
 
-shared_ptr<PhysicsObjects> NanoEvent::GetSegmentMatchedMuons(int minMatches)
+shared_ptr<PhysicsObjects> NanoEvent::GetSegmentMatchedMuons(float minMatchRatio)
 {
   auto looseMuons = GetCollection("LooseMuons");
   auto looseDsaMuons = GetCollection("LooseDSAMuons");
 
-  bool useMatchRatio = false;
-  if(minMatches == 0) useMatchRatio=true;
-
   auto allMuons = make_shared<PhysicsObjects>();
   for(auto muon : *looseMuons){
     allMuons->push_back(muon);
   }
   for(auto dsaMuon : *looseDsaMuons){
 
-    if(useMatchRatio) {
-      float nHits = float(dsaMuon->Get("trkNumDTHits")) + float(dsaMuon->Get("trkNumCSCHits"));
-      minMatches = static_cast<int>(nHits * 2/3);
-    }
+    float nHits = float(dsaMuon->Get("nSegmentHits"));
     
     bool matchFound = false;
-    if(float(dsaMuon->Get("muonMatch1")) >= minMatches) {
-      if(IndexExist(looseMuons, dsaMuon->Get("muonMatch1idx")),true) matchFound = true;
+    float ratio_tmp = float(dsaMuon->Get("muonMatch1")) / nHits;
+    if(ratio_tmp >= minMatchRatio) {
+      if(MuonIndexExist(looseMuons, dsaMuon->Get("muonMatch1idx")),true) matchFound = true;
     }
-    if(!matchFound && float(dsaMuon->Get("muonMatch2")) >= minMatches) {
-      if(IndexExist(looseMuons, dsaMuon->Get("muonMatch2idx")),true) matchFound = true;
+    ratio_tmp = float(dsaMuon->Get("muonMatch2")) / nHits;
+    if(!matchFound && ratio_tmp >= minMatchRatio) {
+      if(MuonIndexExist(looseMuons, dsaMuon->Get("muonMatch2idx")),true) matchFound = true;
     }
-    if(!matchFound && float(dsaMuon->Get("muonMatch3")) >= minMatches) {
-      if(IndexExist(looseMuons, dsaMuon->Get("muonMatch3idx")),true) matchFound = true;
+    ratio_tmp = float(dsaMuon->Get("muonMatch3")) / nHits;
+    if(!matchFound && ratio_tmp >= minMatchRatio) {
+      if(MuonIndexExist(looseMuons, dsaMuon->Get("muonMatch3idx")),true) matchFound = true;
     }
-    if(!matchFound && float(dsaMuon->Get("muonMatch4")) >= minMatches) {
-      if(IndexExist(looseMuons, dsaMuon->Get("muonMatch4idx")),true) matchFound = true;
+    ratio_tmp = float(dsaMuon->Get("muonMatch4")) / nHits;
+    if(!matchFound && ratio_tmp >= minMatchRatio) {
+      if(MuonIndexExist(looseMuons, dsaMuon->Get("muonMatch4idx")),true) matchFound = true;
     }
-    if(!matchFound && float(dsaMuon->Get("muonMatch5")) >= minMatches) {
-      if(IndexExist(looseMuons, dsaMuon->Get("muonMatch5idx")),true) matchFound = true;
+    ratio_tmp = float(dsaMuon->Get("muonMatch5")) / nHits;
+    if(!matchFound && ratio_tmp >= minMatchRatio) {
+      if(MuonIndexExist(looseMuons, dsaMuon->Get("muonMatch5idx")),true) matchFound = true;
```

Comment:
Can we make it a loop, e.g. like this:

```C++
for(int i=1; i<=5; i++){
    float ratio_tmp = float(dsaMuon->GetMuonMatch(i)) / nHits;
    if(ratio_tmp < minMatchRatio) continue;
    matchFound |= MuonIndexExist(looseMuons, dsaMuon->GetMuonMatchIndex(i)), true);
}
```

Reply:
Updated, names of new functions: GetMatchesForNBestMatch(int N) and GetMatchIdxForNBestMatch(int N), as it's for the 1st - 5th best matches. Let me know if you have better suggestions